### ISSUE TITLE
SVD shuffling exp.

### ISSUE THREAD

**Issue by anhhuyalex** [anhhuyalex/renormalization#7]
# principle component with decorrelated
# shuffle the values of principal components: 
# each sample now has its 
instead of zeroing them out, independently shuffle across all data points, keep the top k dimensions
make a plot of no. of dimensions kept
<img width="1179" alt="image" src="https://github.com/anhhuyalex/renormalization/assets/17817335/d92d1803-0d9f-4a9b-a51f-d694ed2c3894">

**Comment by anhhuyalex** [comment]
try zeroing out instead of shuffling

**Comment by anhhuyalex** [comment]
Rerun multiple seeds

**Comment by anhhuyalex** [comment]
Retrain on PCA transformed space instead of original space

**Comment by anhhuyalex** [comment]
Train on fashion-mnist

**Comment by anhhuyalex** [comment]
Match iid gaussian values that match mnist

**Comment by anhhuyalex** [comment]
try synthetic datasets 

try no inverse transform experiment with different regularization
add when N=5000 or even lower 

try when N<D: 

fix D and vary N and look at training curves
fix D and vary N look at train dynamics for different N: 
plot on X axis epoch or iteration: 
plot train / test loss on a log scale

try physical coarse-graining

**Comment by anhhuyalex** [comment]
no inverse transform 
vary lrs 0.0 1e-10 1e-8 1e-5 1e-2
for k in 1 3 5 7 9 11 13 15 17; do python -u mnist_classification_shuffled_svd.py --data mnist --fileprefix Adamshuffledsvd_highsignal_no_regularize_multiclass_mnist_MAR16 --optimizer_type adam --lr 0.001 --lr_scheduler OneCycleLR --multiclass_lr   -b 500  --num_train_samples $j --epochs 2000  --wd $1 --is_high_signal_to_noise True --is_shuffle_signal True --is_inverse_transform False --highsignal_pca_components_kept $i --save_dir /scratch/qanguyen/imagenet_info; done

do inverse transform 
vary lrs 0.0 1e-10 1e-8 1e-5 1e-2
for k in 1 3 5 7 9 11 13 15 17; do python -u mnist_classification_shuffled_svd.py --data mnist --fileprefix Adamshuffledsvd_highsignal_no_regularize_multiclass_mnist_doinversetransform_MAR16 --optimizer_type adam --lr 0.001 --lr_scheduler OneCycleLR --multiclass_lr   -b 500  --num_train_samples $j --epochs 2000  --wd $1 --is_high_signal_to_noise True --is_shuffle_signal True --is_inverse_transform True --highsignal_pca_components_kept $i --save_dir /scratch/qanguyen/imagenet_info; done

**Comment by anhhuyalex** [comment]
for k in 1 3 5 7 9 11 13 15 17; do python -u mnist_classification_shuffled_svd.py --data randomfeaturesmnist --randomfeatures_target_size 14 --fileprefix Adamshuffledsvd_highsignal_no_regularize_randomfeatures14_mnist_noinversetransform_MAR27_wd_$1 --optimizer_type adam --lr 0.001 --lr_scheduler OneCycleLR --multiclass_lr   -b 500  --num_train_samples $j --epochs 2000  --wd $1 --is_high_signal_to_noise True --is_shuffle_signal True --is_inverse_transform False --highsignal_pca_components_kept $i --save_dir /scratch/qanguyen/imagenet_info; done

for k in 1 3 5 7 9 11 13 15 17; do python -u mnist_classification_shuffled_svd.py --data randomfeaturesmnist --randomfeatures_target_size 7 --fileprefix Adamshuffledsvd_highsignal_no_regularize_randomfeatures7_mnist_noinversetransform_MAR27_wd_$1 --optimizer_type adam --lr 0.001 --lr_scheduler OneCycleLR --multiclass_lr   -b 500  --num_train_samples $j --epochs 2000  --wd $1 --is_high_signal_to_noise True --is_shuffle_signal True --is_inverse_transform False --highsignal_pca_components_kept $i --save_dir /scratch/qanguyen/imagenet_info; done

**Comment by anhhuyalex** [comment]
Try scaling epochs to equilibrate # of iterations
Try other kinds of noise (e.g. instead of shuffling try Gaussian noise ?)
Try plotting the norm of the model
simplify to binary classification (e.g. predict evens vs. odds, or predict 0s vs 1s)
Plot norm of difference between learned weight at full N as we fit the model with fewer N

**Comment by anhhuyalex** [comment]
for k in 1 3 5 7 9 11 13 15 17; do python -u mnist_classification_shuffled_svd.py --data mnist --randomfeatures_target_size 7 --fileprefix Adamshuffledsvd_highsignal_no_regularize_mnist_noinversetransform_APR7_wd_$1 --optimizer_type adam --lr 0.001 --lr_scheduler OneCycleLR --multiclass_lr   -b 500  --num_train_samples $j --epochs 2000  --wd $1 --is_high_signal_to_noise True --is_shuffle_signal Gaussian --is_inverse_transform False --highsignal_pca_components_kept $i --save_dir /scratch/qanguyen/imagenet_info; done

for k in 1 3 5 7 9 11 13 15 17; do python -u mnist_classification_shuffled_svd.py --data mnist --randomfeatures_target_size 7 --fileprefix Adamshuffledsvd_highsignal_no_regularize_mnist_doinversetransform_APR7_wd_$1 --optimizer_type adam --lr 0.001 --lr_scheduler OneCycleLR --multiclass_lr   -b 500  --num_train_samples $j --epochs 2000  --wd $1 --is_high_signal_to_noise True --is_shuffle_signal Gaussian --is_inverse_transform True --highsignal_pca_components_kept $i --save_dir /scratch/qanguyen/imagenet_info; done

**Comment by anhhuyalex** [comment]
retrain things with lbfgs
why is difference from reference weights: distance not 0 for D=1.0
train reference weights on the same data
do \frac { } { }tional change of reference weights: delta W / norm (W_reference) instead of just delta W

check optimization: the train directly on PCA & train on inverse PCA
Gaussian experiment: for each principal component, compute each mean and variance, and then sample from that Gaussian mean and variance instead of N(0,1)

**Comment by anhhuyalex** [comment]
rerun N=60k with a lot more iterations for lbfgs, look at using tolerance to stop optimizer instead of max_iter
try initializing lbfgs from zero: plot difference from random-init lbfgs solution and zero-init solution
check if my lbfgs has converged
why do we see a drop in accuracy between d=0.2 and d=0.4
at fixed levels of coarse graining, as we vary N, how is the model changing (plot the norm of the difference in weights vs. N)

rerun gaussian exp. with adam

**Comment by anhhuyalex** [comment]
write down loss function explicitly ?

**Comment by anhhuyalex** [comment]
for k in 1 3 5 7 9 11 13 15 17; do python -u mnist_classification_shuffled_svd.py --data mnist --fileprefix sklearnlbfgs_gaussianshuffledsvd_highsignal_mnist_noinversetransform_zeroinit_MAY25_wd_$1 --optimizer_type sklearn_lbfgs --is_random_init False --lr 0.001 --lr_scheduler OneCycleLR --multiclass_lr   -b 500  --num_train_samples $j  --wd $1 --is_high_signal_to_noise True --is_shuffle_signal True --is_inverse_transform False --highsignal_pca_components_kept $i --save_dir /scratch/qanguyen/imagenet_info; done

**Comment by anhhuyalex** [comment]
doinversetransform / noinversetransform
wd=[0.0,1e-2,1.0]
for k in 1 3 5; do python -u mnist_classification_shuffled_regression.py --data mnist --fileprefix sklearnlbfgs_shuffledregression_highsignal_mnist_doinversetransform_zeroinit_JUN2_wd_$1 --optimizer_type sklearn_lbfgs --lr 0.001 --lr_scheduler OneCycleLR --multiclass_lr   -b 500  --num_train_samples $j  --wd $1 --is_high_signal_to_noise True --is_shuffle_signal True --is_inverse_transform True --highsignal_pca_components_kept $i --true_noise 1.0 --save_dir /scratch/qanguyen/imagenet_info; done # 5 7 9 11 13 15 17

**Comment by anhhuyalex** [comment]
shuffle every row independently
doinversetransform / noinversetransform
wd=[0.0,1e-2,1.0]
for i in {0..10}; do WANDB_MODE=offline ./l2l/bin/python3 -u gmm_transformer_labels.py --data ./cache --fileprefix transformer1layer_lr_0.01_no_posenc --position_encoding False --SLURM_ARRAY_TASK_ID $SLURM_ARRAY_TASK_ID --batch-size 128 --optimizer SGD --lr 0.01 --wd $1 --epsilon 0.0 --burstiness 1 --init_scale_qkv 1.0 --scale_target 1.0 --is_all_tasks_seen True --L 2  --epochs 500 --arch causal_transformer_embed --is_train_random_length_seqs False --len_context 100 --no-wandb_log --wandb_project l2l --is_temperature_fixed False --wandb_group_name gmm_jun30 ; done # matchgautamclassifier_

**Comment by anhhuyalex** [comment]
run experiment where shuffle both train and test sets using the same permutation
hypothesis: train and test loss should be the same

**Comment by anhhuyalex** [comment]
If we shuffle high 
Shuffle high variance explained / low frequency components: we should not see cross over, more data should always help
What happens when we shuffle random components 
Train model just on non-shuffled data and test on shuffled data (where all columns are shuffled together)
Do this and make sure we still get cross over phenomenon
Track down adversarial robustness bugs vs. features literature 
Theory on out of distribution error
Potentially look for this in a multilayer perceptron 
How correlated are the PCA dimensions with the label, vs. size of singular value (i.e. strength of signal) might be two separate problems
Send the sanity checks + literature review

**Comment by anhhuyalex** [comment]
TO DOs: 
- get linear regression results for shuffling, wave settings

**Comment by anhhuyalex** [comment]
Things to put in paper
- analytics for linear regression
- logistic regression, models with nonlinearities empirics
when we shuffle high frequency components, we get non monotonicity
when we shuffle low frequency components, things just kind of break 
- what happens when we get a lot lot of data

when train dist = test dist, more data always helps (modulo interpolating regime)
when train dist is far away from test dist, more data can hurt (shuffle weak signal)

covariate shift assumption: fix p(y|x), change p(x)
we are actually changing both p(x) and p(y|x)
=> we see different phenomena here

**Comment by anhhuyalex** [comment]
Try svd shuffling setting with CNNs
